### DISCUSSION TITLE
Handling native dependencies

### DISCUSSION THREAD

**Discussion by mipastgt** [shannah/jdeploy#88] [General]
After reading all the documentation it is still not clear to me whether jdeploy can handle native dependencies. I have several jar files in my project which contain native shared libraries for one particular platform. E.g. something like mylib-macos-x64.jar. Putting all this into one large uber-jar would make it extremely big. Also my build framework already creates a platform specific  uber-jar which works and I would just like to use that with jdeploy. Is that possible and if yes how?

**Comment by Birdasaur** [comment]
THIs sounds very much like one of my use cases that i want to consider jDEploy to. (nasa worldwind app that has native jogl jars etc)
BUmping this and would love if someone whom knows more about jdeploy could shed some light on it.

**Reply by shannah** [reply]
I have toyed with a few solutions for this.  One is using maven dependencies that download at install-time.  This works, but requires the dependencies to be available on maven central (or other publicly accessible repo).  I had a few example projects with this setup but am struggling to find them now.

Another thought was to allow you to have separate bundles for each platform, and have the app download the appropriate one at runtime.  

Do you have a project that I can experiment with that needs this?

**Reply by Birdasaur** [reply]
Unfortunately, the source is currently not open, though my organization is working towards open sourcing the application.
If I had to guess right now in a naive manner as to which one of the above approaches I would prefer... I'd actually prefer the latter.
While having separate bundles seems like a hassle at first it would make the resulting deployment more "atomic" and reduce the risk and complexity of failure modes. 
For this application I am currently using an old school Maven Assembly plugin to build a Fat Jar then packaging a JRE in one mighty zip archive. (haven't had a chance to put together a Gradle Jpackage script yet)
It works though I am experiencing the pain of folks "not having a link to just click". Also I can only build the app on my OS (which happens to be Windows) and I have to rely on others to build it for Mac and Linux. (yes we have interested users for those as well.)

Looking at JDeploy I feel like it mostly addresses both of these issues and mostly shuts up the web app fanboys.

**Reply by shannah** [reply]
> Also I can only build the app on my OS (which happens to be Windows) and I have to rely on others to build it for Mac and Linux. 

Can you build the jar files for those platforms on Windows, or do you also need to be using Mac/Linux for that?

**Reply by shannah** [reply]
If you're willing to be a guinea pig I can have a poc ready to do separate bundles by platform by next week.

Do you plan to distribute by npm or github releases?

**Reply by mipastgt** [reply]
In the end I would like to host the installers myself but for initial testing npm might also be ok. My free GitHub account may be a bit too space limited.

**Reply by shannah** [reply]
Github has no space limits on releases even for free accounts.  Other than the 2gb per file, which shouldn't be an issue.

I would like to add self hosting support but it hasn't reached top priority yet, as there isn't a clear use case for it that adds any advantages over using npm or github.

The primary reason to use github releases would be to integrate into github actions.  Eg you can do things like distribute a version that stays in sync with a specific branch.

**Reply by mipastgt** [reply]
Then we can also use GitHub releases if you can provide me with some description of the setup.

**Comment by mipastgt** [comment]
I agree with what @Birdasaur said. I'd also prefer to be able to create separate artifacts for each platform. I'm primarily interested in Jetbrains Compose Multiplatform desktop apps and the built-in builder already has this structure. If you need an example project for that you can just generate one with a button click here: https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-multiplatform-create-first-app.html#create-a-project-using-the-wizard 

@Birdasaur for the OS problem you mentioned above I just use GitHub actions. I build locally on a Mac and build the Windows and Linux variants via a GitHub action on their provided runners, using their free plan.

**Reply by shannah** [reply]
@mipastgt Do compose multiplatform desktop apps contain a lot of heavy native libs that benefit from creating separate targeted bundles for each platform (win/mac/linux), or do you generally distribute the same jars to every platform?

**Reply by mipastgt** [reply]
Oh yes. They contain by default the native Skia library which is several MB and they are bundled with the platform JRE.

**Comment by shannah** [comment]
I have released v5.1.1 that adds support for platform-specific bundles, so that you can choose which native libraries get included for each platform.

https://github.com/shannah/jdeploy/releases/tag/v5.1.1
